README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# edbuildmapr

<!-- badges: start -->

<!-- badges: end -->

The goal of `edbuildmapr` is to simplify the use of school district
shapefiles. The package import US Census Bureau, Education Demographic
and Geographic Estimates Program, Composite School District Boundaries
Files for 2013-2018 with the option to attach EdBuild’s master dataset
of school district finance, student demographics, and community economic
indicators for every school district in the United States. The master
dataset is built from the US Census, Annual Survey of School System
Finances (F33) and joins data from the National Center for Education
Statistics, Common Core of Data; the US Census, Small Area Income and
Poverty Estimates; and the US Census, Education Demographic and
Geographic Estimates. Additional functions in the package create a
dataset of all school district neighbors as either a dataframe or a
shapefile and maps of selected districts at the state or neighbor level
symbolized by a selected variable in EdBuild’s master dataset. For full
details about EdBuild’s data processing please see EdBuild (2020)
[Methodology](https://data.edbuild.org/).

## Installation

You can install the released version of edbuildmapr from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("edbuildmapr")
```

## Example

To import school district spatial data for all districts with EdBuild’s
master data:

``` r
library(edbuildmapr)
## import school district shapefile for the year 2018 with EdBuild's master dataset attached. 
sd_shp_18 <- sd_shapepull("2018", with_data = TRUE)
#> Reading layer `shapefile_1718_4269' from data source `/private/var/folders/5n/cz5rxt2d02jcqlt7q25sy37m0000gp/T/RtmpHoTFqD/filea47572b1be98/shapefile_1718_4269.shp' using driver `ESRI Shapefile'
#> Simple feature collection with 13330 features and 6 fields
#> geometry type:  MULTIPOLYGON
#> dimension:      XY
#> bbox:           xmin: -179.1686 ymin: 18.91382 xmax: 179.7487 ymax: 71.38881
#> CRS:            4269
#> NOTE::This shapefile has been simplified to make analysis quicker. For final vizualizations, please use the unsimplified shapefiles available through NCES.

str(sd_shp_18)
#> Classes 'sf' and 'data.frame':   13309 obs. of  47 variables:
#>  $ FIPS       : chr  "04" "04" "04" "04" ...
#>  $ GEOID      : chr  "0407530" "0401740" "0401160" "0400720" ...
#>  $ NAME       : chr  "Santa Cruz Valley Union High School District" "Casa Grande Union High School District" "Bicentennial Union High School District" "Antelope Union High School District" ...
#>  $ sdType     : chr  "secon" "secon" "secon" "secon" ...
#>  $ State      : chr  "Arizona" "Arizona" "Arizona" "Arizona" ...
#>  $ Postal     : chr  "AZ" "AZ" "AZ" "AZ" ...
#>  $ CONUM      : int  4021 4021 4012 4027 4027 4015 4013 4013 4013 4013 ...
#>  $ ENROLL     : int  367 3376 111 220 10959 1842 7780 4069 15029 27268 ...
#>  $ LRPP       : num  11275 5699 14189 11650 3136 ...
#>  $ SRPP       : num  2659 4295 1477 1195 4468 ...
#>  $ SLRPP      : num  13935 9994 15667 12845 7604 ...
#>  $ LR         : num  4138000 19241000 1575000 2563000 34370000 ...
#>  $ SR         : num  976000 14500000 164000 263000 48967000 ...
#>  $ SLR        : num  5114000 33741000 1739000 2826000 83337000 ...
#>  $ COLIn      : num  0.971 0.971 1.002 0.957 0.957 ...
#>  $ SRPP_cola  : num  2739 4423 1475 1249 4669 ...
#>  $ LRPP_cola  : num  11612 5870 14161 12173 3277 ...
#>  $ SLRPP_cola : num  14351 10293 15635 13423 7946 ...
#>  $ County     : chr  "Pinal County" "Pinal County" "La Paz County" "Yuma County" ...
#>  $ dType      : chr  "1-Regular local school district that is NOT a component of a supervisory union" "1-Regular local school district that is NOT a component of a supervisory union" "1-Regular local school district that is NOT a component of a supervisory union" "1-Regular local school district that is NOT a component of a supervisory union" ...
#>  $ urb        : chr  "31-Town: Fringe" "23-Suburb: Small" "43-Rural: Remote" "42-Rural: Distant" ...
#>  $ op_shcls   : int  2 2 1 2 7 3 5 4 11 17 ...
#>  $ enroll_ccd : int  367 3376 111 220 10959 1842 7780 4069 15029 27268 ...
#>  $ dFRL       : int  NA 2025 83 174 7627 1021 2938 NA 8672 22555 ...
#>  $ dLEP       : int  25 97 11 14 1234 26 129 126 809 2546 ...
#>  $ dWhite     : int  25 815 38 57 1333 958 2670 1410 4106 1223 ...
#>  $ dBlack     : int  31 123 0 0 133 32 798 291 1245 2288 ...
#>  $ dHispanic  : int  301 2043 72 162 9264 743 3699 2182 8606 22223 ...
#>  $ dAsian_PI  : int  1 50 0 0 56 16 246 37 487 456 ...
#>  $ dHaw_PI    : int  0 9 0 0 15 0 40 20 34 36 ...
#>  $ dAmInd_AK  : int  8 254 0 1 59 40 56 40 370 625 ...
#>  $ d2races    : int  1 82 1 0 99 53 271 89 181 417 ...
#>  $ pctNW      : num  0.932 0.759 0.658 0.741 0.878 ...
#>  $ TPop       : int  21741 103199 11246 6909 205219 67376 139522 76028 373568 754070 ...
#>  $ StPop      : int  747 5585 311 357 12121 2661 8131 4493 20614 44227 ...
#>  $ StPov      : int  157 1069 31 63 3167 609 817 530 4610 11530 ...
#>  $ StPovRate  : num  0.2102 0.1914 0.0997 0.1765 0.2613 ...
#>  $ MPV        : int  115800 126000 70500 80300 123300 137900 265300 198600 172100 169400 ...
#>  $ MHI        : int  44167 45789 30432 46047 43988 41673 75226 63832 44856 43643 ...
#>  $ sd_area    : num  718 1300 3809 3065 2454 ...
#>  $ st_per_sqmi: num  0.5109 2.5973 0.0291 0.0718 4.4654 ...
#>  $ sd_type    : chr  "secon" "secon" "secon" "secon" ...
#>  $ state_id   : chr  "AZ-4454" "AZ-4453" "AZ-4515" "AZ-4506" ...
#>  $ dIEP       : int  55 494 21 37 964 215 812 499 1875 3042 ...
#>  $ year       : chr  "2018" "2018" "2018" "2018" ...
#>  $ geometry   :sfc_MULTIPOLYGON of length 13309; first list element: List of 1
#>   ..$ :List of 1
#>   .. ..$ : num [1:135, 1:2] -111 -111 -111 -111 -111 ...
#>   ..- attr(*, "class")= chr  "XY" "MULTIPOLYGON" "sfg"
#>  $ frl_rate   : num  NA 0.6 0.748 0.791 0.696 ...
#>  - attr(*, "sf_column")= chr "geometry"
#>  - attr(*, "agr")= Factor w/ 3 levels "constant","aggregate",..: NA NA NA NA NA NA NA NA NA NA ...
#>   ..- attr(*, "names")= chr  "FIPS" "GEOID" "NAME" "sdType" ...
```

## Authors

  - Megan Brodzik (<meganbrodzik@edbuild.org>), maintainer
  - Cecilia Depman (<cecilia.depman@gmail.com>), author
  - Sara Hodges (<sara.hodges@gmail.com>), author


NEWS.md
# edbuildmapr 0.3.0

* added parameter to `<sd_map()>` and `<sd_neighbor_map()>` to select the data year to map
* updated all functions with 2019 data

# edbuildmapr 0.2.0

* added new function `<state_shapepull()>` to import state shapefile that matches school district borders
* added parameter to `<sd_map()>` to map a county instead of a state
* corrected legend display size in `<sd_map()>` and `<sd_neighbor_map()>`
* updated all functions with 2018 data

# edbuildmapr 0.1.1

* Changed import of `st_make_valid` from `lwgeom` to `sf` as object `st_make_valid` is exported by `sf` version 0.9.1.

# edbuildmapr 0.1.0

* Initial release of `edbuildmapr` to automate the downloading and analysis of school district spatial data. 


R/sd_map.R
#' @title A function to create a map of all school districts in a state
#' @name sd_map
#' @description This function allows you to create a map of all school
#'   districts, in each state in the United States, symbolized by selected
#'   variables from the EdBuild master dataset.
#' @param state The state for which you want to map school districts.  Defaults
#'   to New Jersey.
#' @param county The county for which you want to map school districts.
#'   Defaults to NULL. To view a full list of counties use
#'   \code{sd_shapepull(year = "2018", with_data = TRUE)}
#' @param map_var Variable by which to symbolize the map. \itemize{
#'   \item{\code{Student Poverty} colors by student poverty rate}
#'   \item{\code{Total Revenue} colors by state and local revenue per pupil}
#'   \item{\code{Local Revenue} colors by local revenue per pupil} \item
#'   {\code{State Revenue} colors by state revenue per pupil} \item
#'   {\code{Percent Nonwhite} colors by percent nonwhite enrollment} \item
#'   {\code{Median Household Income} colors by median household income } \item{
#'   \code{Median Property Value} colors by owner-occupied median property value
#'   } \item {\code{FRL} colors by free and reduced price lunch rate} } Defaults
#'   to \code{Student Poverty}
#' @param level Selects which level of school districts you want displayed in
#'   the map. \itemize{ \item{\code{elem} {displays elementary and unified
#'   districts}} \item{\code{secon} {displays secondary and unified districts}}}
#'   Defaults to \code{elem}.
#' @param legend If TRUE, legend is visible. Defaults to TRUE.
#' @keywords school districts map EdBuild
#' @usage sd_map(state="New Jersey", county = NULL, map_var = "Student Poverty",
#'   level = "elem", legend= TRUE)
#' @import dplyr sf magrittr
#' @importFrom tmap tm_shape tm_fill tm_borders tm_layout
#' @return An image of the map which can be written out with
#'   \code{tmap::tmap_save(map, '~/Documents/map.png')}
#' @seealso \code{\link{sd_neighbor_map}}
#' @export
#' @examples
#' \donttest{map <- sd_map(state="Georgia", map_var = "Percent Nonwhite",
#'  level = "elem", legend= TRUE)}


sd_map <- function (state="New Jersey", county = NULL,  map_var = "Student Poverty", level = "elem", legend= TRUE) {
  shape <- sd_shapepull(data_year = "2018", with_data = TRUE)

  states <- shape
  sf::st_geometry(states) <- NULL

  states_list <- as.list(levels(as.factor(states$State)))

  county_list <- as.list(levels(as.factor(states$County)))

  state_shape <- shape %>%
    dplyr::mutate(NAME = as.character(NAME),
                  FIPS = as.character(FIPS),
                  State = as.character(State),
                  Postal = as.character(Postal),
                  sdType= as.character(sdType)) %>%
    dplyr::filter(State == state) %>%
    dplyr::mutate(frl_rate = dplyr::case_when(frl_rate > 1 ~  NA_real_,
                                              TRUE ~ frl_rate),
                  MPV = as.numeric(MPV),
                  MHI = as.numeric(MHI))

  pos_vars <- list("Student Poverty", "Percent Nonwhite", "FRL", "Total Revenue", "Local Revenue",
                   "State Revenue", "Median Household Income", "Median Property Value")

  shape.clean <- sf::st_make_valid(state_shape) # making all geometries valid

  if(map_var %in% pos_vars == FALSE) {

    message("Error: School district mapping is only available for the following variables: Student Poverty, Percent Nonwhite, FRL, Total Revenue, Local Revenue, State Revenue, Median Household Income and Median Property Value")
  }

  else if (state %in% states_list == FALSE) {
    message("Error: Please check your spelling of the state.")
  }

  else if (level != "elem" & level != "secon") {
    message("Error: Please select the school district level you would like displayed.
              'elem' displays elementary and unifed districts.
              'secon' displays secondary and unified distrcts.")
  }
  else {
    if(map_var == "Student Poverty"){
      colors <- c('#dff3fe', '#92DCF0', '#49B4D6', '#2586a5', '#19596d')
      title_name = "Student Poverty"
      variable = "StPovRate"
      breaks = c(0, .1, .2, .3, .4, 1)
      legend_format = list(fun=function(x) paste0(formatC(x*100, digits=0, format="f"), " %"))
      format_color = '#2586a5'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/student_poverty_rate_map.png')")
    }
    else if(map_var == "Total Revenue"){
      colors <- c('#edf8fb', '#b2e2e2', '#66c2a4', '#277f4d', '#1a5f38', '#0f3720', '#001107')
      title_name = "Total Revenue Per Pupil"
      variable = "SLRPP"
      breaks = c(0, 7500, 10000, 12500, 15000, 20000, 100000, 1000000)
      #legend_format = list(fun=function(x) paste0("$ ", formatC(x, digits=0, format="f", big.mark = ",")))
      legend_format = list(fun = function(x) paste0("$", x/1000, "k"))
      format_color = '#277f4d'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/total_revenue_pp_map.png')")

    }
    else if(map_var == "Local Revenue"){
      colors <- c('#e5f5f9', '#ccece6', '#99d8c9', '#66c2a4', '#41ae76', '#238b45', '#006d2c', '#00441b', '#001107')
      title_name = "Local Revenue Per Pupil"
      variable = "LRPP"
      breaks = c(0, 1000, 2500, 5000, 7500, 10000, 15000, 25000, 100000, 1000000)
      # legend_format = list(fun=function(x) paste0("$ ", formatC(x, digits=0, format="f", big.mark = ",")))
      legend_format = list(fun = function(x) paste0("$", x/1000, "k"))
      format_color = '#277f4d'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/local_revenue_pp_map.png')")

    }
    else if(map_var == "State Revenue"){
      colors <- c('#e5f5f9', '#ccece6', '#99d8c9', '#66c2a4', '#41ae76', '#238b45', '#006d2c', '#00441b', '#001107')
      title_name = "State Revenue Per Pupil"
      variable = "SRPP"
      breaks = c(0, 1000, 2500, 5000, 7500, 10000, 15000, 25000, 100000, 1000000)
      # legend_format = list(fun=function(x) paste0("$ ", formatC(x, digits=0, format="f", big.mark = ",")))
      legend_format = list(fun = function(x) paste0("$", x/1000, "k"))
      format_color = '#277f4d'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/state_revenue_pp_map.png')")

    }
    else if(map_var == "Median Household Income"){
      colors <- c('#d73027', '#e55d30', '#f28648', '#f9b761', '#ffffbf','#c6e7ef', '#73b9d1', '#4483ba', '#275ea5')
      title_name = "Median Household Income"
      variable = "MHI"
      breaks = c(0, 15000, 25000, 40000, 50000, 65000, 85000, 100000, 150000, 2000001)
      # legend_format = list(fun=function(x) paste0("$ ", formatC(x, digits=0, format="f", big.mark = ",")))
      legend_format = list(fun = function(x) paste0("$", x/1000, "k"))
      format_color = '#73b9d1'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/median_household_income_map.png')")

    }
    else if(map_var == "Median Property Value"){
      colors <- c('#bf812d', '#dfc27d', '#f6e8c3', '#f5f5f5', '#c7eae5', '#80cdc1', '#35978f', '#01665e')
      title_name = "Median Property Value"
      variable = "MPV"
      breaks = c(0, 50000, 100000, 150000, 200000, 250000, 300000, 350000, 400000, 2000001)
      # legend_format = list(fun=function(x) paste0("$ ", formatC(x, digits=0, format="f", big.mark = ",")))
      legend_format = list(fun = function(x) paste0("$", x/1000, "k"))
      format_color = '#80cdc1'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/median_property_value_map.png')")

      }
    else if(map_var == "FRL"){
      colors <- c('#e3b1a5', '#ce7965',  '#c2573e', '#823a29', '#4e2319')
      title_name = "FRL Rate"
      variable = "frl_rate"
      breaks = c(0, .2, .4, .6, .8, 1)
      legend_format = list(fun=function(x) paste0(formatC(x*100, digits=0, format="f"), " %"))
      format_color = '#823a29'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/frl_rate_map.png')")

    }
    else if(map_var == "Percent Nonwhite"){
      colors <- c('#f2f0f7', '#cbc9e2', '#9c7dc4', '#7d52b7', '#54278f')
      title_name = "Percent Nonwhite"
      variable = "pctNW"
      breaks = c(0, .2, .4, .6, .8, 1)
      legend_format = list(fun=function(x) paste0(formatC(x*100, digits=0, format="f"), " %"))
      format_color = '#7d52b7'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/percent_nonwhite_map.png')")

    }
    else {
      message("Error: School district mapping is only available for the following variables: Student Poverty, Percent Nonwhite, FRL, Total Revenue, Local Revenue, State Revenue, Median Household Income and Median Property Value")
    }

    if(level == "elem") {
      shape.clean <- shape.clean %>%
        dplyr::filter(sdType == "elem" | sdType == "uni")
    }

    else if(level == "secon") {
      shape.clean <- shape.clean %>%
        dplyr::filter(sdType == "secon" | sdType == "uni")
    }

    if(!is.null(county)) {

      if (county %in% county_list == FALSE) {
        message("Error: Please check your spelling of the county. To view correct county spellings use sd_shapepull(with_data = TRUE)")
      }
      else if (county %in% county_list == TRUE){

        county_shape <- shape.clean %>%
        dplyr::filter(County == county)

      map <- tmap::tm_shape(county_shape) +
        tmap::tm_fill(variable, breaks=breaks, title = title_name,
                      palette = colors,
                      legend.format=legend_format) +
        tmap::tm_shape(county_shape) +
        tmap::tm_borders(lwd=.15, col = "#cfcccc", alpha = 1) +
        tmap::tm_layout(bg.color = NA,
                        main.title = paste0(county, ", ", state),
                        main.title.position = "center",
                        main.title.color = format_color,
                        legend.position = c("left", "bottom"),
                        legend.text.color = format_color,
                        legend.title.color = format_color,
                        frame = FALSE,
                        legend.outside = TRUE,
                        legend.outside.position = "left",
                        legend.outside.size = .5,
                        legend.text.size = .6 ) +
        tmap::tm_layout(legend.show = legend)

      }
    }
    else {

    map <- tmap::tm_shape(shape.clean) +
      tmap::tm_fill(variable, breaks=breaks, title = title_name,
                    palette = colors,
                    legend.format=legend_format) +
      tmap::tm_shape(shape.clean) +
      tmap::tm_borders(lwd=.15, col = "#cfcccc", alpha = 1) +
      tmap::tm_layout(bg.color = NA,
                      main.title = state,
                      main.title.position = "center",
                      main.title.color = format_color,
                     # legend.position = c("left", "bottom"),
                      legend.text.color = format_color,
                      legend.title.color = format_color,
                      frame = FALSE,
                      legend.outside = TRUE,
                      legend.outside.position = "left",
                     legend.outside.size = .5,
                     legend.text.size = .6
                     # outer.margins = c(0,0,0,0)
                     ) +
      tmap::tm_layout(legend.show = legend)
    }
     map  ### view the map
  }
}


R/edbuildmapr.R
#'edbuildmapr: A package to download school district geospatial data, perform spatial
#'analysis, and create formatted exportable maps.
#'
#'Import US Census Bureau, Education Demographic and Geographic Estimates
#'Program, Composite School District Boundaries Files for 2013-2018 with the
#'option to attach EdBuild's master dataset of school district finance, student
#'demographics, and community economic indicators for every school district in
#'the United States. The master dataset is built from the US Census, Annual
#'Survey of School System Finances (F33) and joins data from the National Center
#'for Education Statistics, Common Core of Data; the US Census, Small Area
#'Income and Poverty Estimates; and the US Census, Education Demographic and
#'Geographic Estimates. Additional functions in the package create a dataset of
#'all pairs of school district neighbors as either a dataframe or a shapefile
#'and create formatted maps of selected districts at the state or neighbor
#'level, symbolized by a selected variable in EdBuild's master dataset. For full
#'details about 'EdBuild' data processing please see: EdBuild (2020)
#'\href{https://data.edbuild.org/}{Methodology}.

#'
#'@section edbuildmapr functions: The edbuildmapr functions are: \describe{
#'  \item{\code{\link{borders}}}{Create a dataframe or linestring object of the
#'  borders between neighboring districts.} \item{\code{\link{sd_map}}}{Create a
#'  map of all school districts in any state symbolized by a selected variable from
#'  the EdBuild master dataset.} \item{\code{\link{sd_neighbor_map}}}{Create a
#'  map of any school district with its neighbors symbolized by a selected
#'  variable from the EdBuild master dataset.}
#'  \item{\code{\link{sd_shapepull}}}{Import a simplified version the US Census
#'  Bureau, Education Demographic and Geographic Estimates Program (EDGE),
#'  Composite School District Boundaries File for any school year from 2013 to
#'  2018 with the option to attach EdBuild's master dataset of school district
#'  finance, student demographics, and community economic indicators for every
#'  school district in the US.} \item{\code{\link{state_shapepull}}}{Import a
#'  simplified version of state shapefiles matching the US Census Bureau,
#'  Education Demographic and Geographic Estimates Program (EDGE), Composite
#'  School District Boundaries File for 2018.}}
#'@author \itemize{ \item {Megan Brodzik (megan@edbuild.org), maintainer} \item
#'  {Cecilia Depman (cecilia@edbuild.org), author} \item {Sara Hodges
#'  (sara@edbuild.org), author} }
#'@docType package
#'@name edbuildmapr
NULL
#> NULL


R/sd_neighbor_map.R
#' @title A function to create a map of a school district and its neighbors
#' @name sd_neighbor_map
#' @description This function allows you to create a map of any school
#'   district with its neighbors symbolized by a selected variable.
#' @usage sd_neighbor_map(school_district = NULL,
#'  map_var = "Student Poverty", legend= TRUE, type = "like")
#' @param school_district Seven digit NCESID of the school district. Default is
#'   NULL.
#' @param map_var Variable by which to symbolize the map.
#' \itemize{
#' \item {\code{Student Poverty} colors by student poverty rate}
#'   \item {\code{Total Revenue} colors by state and local revenue per pupil}
#'   \item{ \code{Local Revenue} colors by local revenue per pupil}
#'   \item {\code{State Revenue} colors by state revenue per pupil}
#'   \item {\code{Percent Nonwhite} colors by percent nonwhite enrollment}
#'   \item {\code{Median Household Income} colors by owner-occupied median household income}
#'  \item {\code{Median Property Value} colors by median property value}
#'   \item {\code{FRL} colors by free and reduced price lunch rate}
#'   }
#'   Defaults to \code{Student Poverty}
#' @param legend If TRUE, legend is visible. Defaults to TRUE.
#' @param type Indicate which types of neighbors to return.
#'   Defaults to "like", returning a map of neighbors of the same
#'   district type (unified to unified, elementary to elementary and
#'   secondary to secondary). To view all neighbors use "all". This
#'   becomes important for districts like Chicago which have upwards of 50
#'   neighboring school districts, but only 1 type-like neighbor.
#'   Chicago is a unified district with 1 unified neighbor,
#'   16 secondary neighbors, and 32 elementary neighbors.
#' @keywords neighbors map EdBuild
#' @import dplyr stringr magrittr sf
#' @importFrom tmap tm_shape tm_fill tm_borders tm_layout
#' @return An image of the map which can be written out with
#'   \code{tmap::tmap_save(map, '~/Documents/map.png')}
#' @seealso \code{\link{sd_map}}
#' @export
#' @examples
#' \donttest{map <- sd_neighbor_map(school_district = "2601103", "Percent Nonwhite")}

sd_neighbor_map = function(school_district = NULL, map_var = "Student Poverty", legend= TRUE, type = "like"){

  shape <- sd_shapepull(data_year = "2017", with_data = TRUE)

  ncesid <- shape$GEOID

  pos_vars <- list("Student Poverty", "Percent Nonwhite", "FRL", "Total Revenue", "Local Revenue",
                   "State Revenue", "Median Household Income", "Median Property Value")
  if(map_var %in% pos_vars == FALSE) {

    message("Error: School district mapping is only available for the following variables: Student Poverty, Percent Nonwhite, FRL, Total Revenue, Local Revenue, State Revenue, Median Household Income and Median Property Value")
  }

  else if(is.null(school_district)) {
    message( "To generate a table, specify a school district by its NCESID.
    If you do not know the NCESID of a school district, use sd_shapepull() to search for your district.")
  }

  else if(school_district %in% ncesid == FALSE) {
    message( "The district you specified is not available. To generate a table, specify a school district by its NCESID.
    If you do not know the NCESID of a school district, use sd_shapepull() to search for your district.")
  }

  else  {
    sd_url = "https://s3.amazonaws.com/data.edbuild.org/public/Processed+Data/SD+Types/sd_type_17_18.csv"
    sd_type <- read.csv(file = sd_url, stringsAsFactors = FALSE)

    shape.clean <- sf::st_make_valid(shape) # making all geometries valid

    pairs_url = "https://s3.amazonaws.com/data.edbuild.org/public/Processed+Data/Pairs/pairs_1617.csv"
    pairs <- read.csv(file = pairs_url, stringsAsFactors = FALSE) %>%
      dplyr::left_join(sd_type, by = "GEOID") %>%
      dplyr::rename(sdtype_1 = sdType) %>%
      dplyr::left_join(sd_type, by = c("GEOID.1" = "GEOID")) %>%
      dplyr::rename(sdtype_2 = sdType) %>%
      dplyr::mutate(FIPS_1 = substr(GEOID, 0, 2),
                    FIPS_2 = substr(GEOID.1, 0, 2)) %>%
      dplyr::filter(FIPS_1 == FIPS_2,      ## filtering out pairs where they are across state lines
                    length >= 152.4) %>%     ## filtering out pairs where the border is less than 500 feet
      dplyr::mutate(GEOID = as.character(stringr::str_pad(GEOID, width = 7, pad = "0")),
                    GEOID.1 = as.character(stringr::str_pad(GEOID.1, width = 7, pad = "0"))) %>%
      dplyr::filter(GEOID == school_district)

    if(map_var == "Student Poverty"){
      colors <- c('#dff3fe', '#92DCF0', '#49B4D6', '#2586a5', '#19596d')
      title_name = "Student Poverty"
      variable = "StPovRate"
      breaks = c(0, .1, .2, .3, .4, 1)
      legend_format = list(fun=function(x) paste0(formatC(x*100, digits=0, format="f"), " %"))
      format_color = '#2586a5'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/student_poverty_rate_map.png')")
    }
    else if(map_var == "Total Revenue"){
      colors <- c('#edf8fb', '#b2e2e2', '#66c2a4', '#277f4d', '#1a5f38', '#0f3720', '#001107')
      title_name = "Total Revenue Per Pupil"
      variable = "SLRPP"
      breaks = c(0, 7500, 10000, 12500, 15000, 20000, 100000, 1000000)
      #legend_format = list(fun=function(x) paste0("$ ", formatC(x, digits=0, format="f", big.mark = ",")))
      legend_format = list(fun = function(x) paste0("$", x/1000, "k"))
      format_color = '#277f4d'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/total_revenue_pp_map.png')")

    }
    else if(map_var == "Local Revenue"){
      colors <- c('#e5f5f9', '#ccece6', '#99d8c9', '#66c2a4', '#41ae76', '#238b45', '#006d2c', '#00441b', '#001107')
      title_name = "Local Revenue Per Pupil"
      variable = "LRPP"
      breaks = c(0, 1000, 2500, 5000, 7500, 10000, 15000, 25000, 100000, 1000000)
      #legend_format = list(fun=function(x) paste0("$ ", formatC(x, digits=0, format="f", big.mark = ",")))
      legend_format = list(fun = function(x) paste0("$", x/1000, "k"))
      format_color = '#277f4d'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/local_revenue_pp_map.png')")

    }
    else if(map_var == "State Revenue"){
      colors <- c('#e5f5f9', '#ccece6', '#99d8c9', '#66c2a4', '#41ae76', '#238b45', '#006d2c', '#00441b', '#001107')
      title_name = "State Revenue Per Pupil"
      variable = "SRPP"
      breaks = c(0, 1000, 2500, 5000, 7500, 10000, 15000, 25000, 100000, 1000000)
      #legend_format = list(fun=function(x) paste0("$ ", formatC(x, digits=0, format="f", big.mark = ",")))
      legend_format = list(fun = function(x) paste0("$", x/1000, "k"))
      format_color = '#277f4d'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/state_revenue_pp_map.png')")

    }
    else if(map_var == "Median Household Income"){
      colors <- c('#d73027', '#e55d30', '#f28648', '#f9b761', '#ffffbf','#c6e7ef', '#73b9d1', '#4483ba', '#275ea5')
      title_name = "Median Household Income"
      variable = "MHI"
      breaks = c(0, 15000, 25000, 40000, 50000, 65000, 85000, 100000, 150000, 2000001)
      #legend_format = list(fun=function(x) paste0("$ ", formatC(x, digits=0, format="f", big.mark = ",")))
      legend_format = list(fun = function(x) paste0("$", x/1000, "k"))
      format_color = '#73b9d1'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/median_household_income_map.png')")

    }
    else if(map_var == "Median Property Value"){
      colors <- c('#bf812d', '#dfc27d', '#f6e8c3', '#f5f5f5', '#c7eae5', '#80cdc1', '#35978f', '#01665e')
      title_name = "Median Property Value"
      variable = "MPV"
      breaks = c(0, 50000, 100000, 150000, 200000, 250000, 300000, 350000, 400000, 2000001)
      #legend_format = list(fun=function(x) paste0("$ ", formatC(x, digits=0, format="f", big.mark = ",")))
      legend_format = list(fun = function(x) paste0("$", x/1000, "k"))
      format_color = '#80cdc1'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/median_property_value_map.png')")

    }
    else if(map_var == "FRL"){
      colors <- c('#e3b1a5', '#ce7965',  '#c2573e', '#823a29', '#4e2319')
      title_name = "FRL Rate"
      variable = "frl_rate"
      breaks = c(0, .2, .4, .6, .8, 1)
      legend_format = list(fun=function(x) paste0(formatC(x*100, digits=0, format="f"), " %"))
      format_color = '#823a29'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/frl_rate_map.png')")

    }
    else if(map_var == "Percent Nonwhite"){
      colors <- c('#f2f0f7', '#cbc9e2', '#9c7dc4', '#7d52b7', '#54278f')
      title_name = "Percent Nonwhite"
      variable = "pctNW"
      breaks = c(0, .2, .4, .6, .8, 1)
      legend_format = list(fun=function(x) paste0(formatC(x*100, digits=0, format="f"), " %"))
      format_color = '#7d52b7'

      message("NOTE:: save your map to the desired location using: tmap::tmap_save(map, bg = 'transparent', '~/Documents/percent_nonwhite_map.png')")

    }
    else {
      message("Error: School district mapping is only available for the following variables: Student Poverty, Percent Nonwhite, FRL, Total Revenue, Median Household Income and Median Property Value")
    }

    if(type == "like"){

      pairs <- pairs %>%
        dplyr::filter(sdtype_1 == sdtype_2)

      GEOIDS <- c(pairs$GEOID.1, school_district)

      shape_small <- shape %>%
        dplyr::filter(GEOID %in% GEOIDS) %>%
        mutate(MPV = as.numeric(MPV),
                MHI = as.numeric(MHI))

      selected <- shape_small %>%
        dplyr::filter(GEOID == school_district)

      tmap::tm_shape(shape_small) +
        tmap::tm_fill(variable, breaks=breaks, title = title_name,
                      palette = colors,
                      legend.format=legend_format) +
        tmap::tm_shape(shape_small) +
        tmap::tm_borders(lwd=.5, col = "#000000", alpha = 1) +
        tmap::tm_shape(selected) +
        tmap::tm_borders(lwd=1.5, col = "#000000", alpha = 1) +
        tmap::tm_layout(bg.color = NA,
                        legend.position = c("left", "bottom"),
                        legend.text.color = format_color,
                        legend.title.color = format_color,
                        main.title = paste0(selected$NAME, ", ", selected$Postal),
                        main.title.color = format_color,
                        frame = FALSE,
                        legend.outside = TRUE,
                        legend.outside.position = "left",
                        legend.outside.size = .4,
                        legend.text.size = .5) +
        tmap::tm_layout(legend.show = legend)


    }

    else if (type == "all"){

      GEOIDS <- c(pairs$GEOID.1, school_district)

      shape_small <- shape %>%
        dplyr::filter(GEOID %in% GEOIDS) %>%
        mutate(MPV = as.numeric(MPV),
               MHI = as.numeric(MHI))

      selected <- shape_small %>%
        dplyr::filter(GEOID == school_district)

      tmap::tm_shape(shape_small) +
        tmap::tm_fill(variable, breaks=breaks, title = title_name,
                      palette = colors,
                      legend.format=legend_format) +
        tmap::tm_shape(shape_small) +
        tmap::tm_borders(lwd=.5, col = "#000000", alpha = 1) +
        tmap::tm_shape(selected) +
        tmap::tm_borders(lwd=1.5, col = "#000000", alpha = 1) +
        tmap::tm_layout(bg.color = NA,
                        legend.position = c("left", "bottom"),
                        legend.text.color = format_color,
                        legend.title.color = format_color,
                        main.title = paste0(selected$NAME, ", ", selected$Postal),
                        main.title.color = format_color,
                        frame = FALSE,
                        legend.outside = TRUE,
                        legend.outside.position = "left",
                        legend.outside.size = .4,
                        legend.text.size = .5) +
        tmap::tm_layout(legend.show = legend)
    }
  }
}
